### PR TITLE
Vcert-java initial implementation

### PR DESCRIPTION
**Pull request by pegerto**

### CODE REVIEW

**Comment by pegerto** [comment]
Hi @tr1ck3r 

could you take a look at the current progress,  we are progressing on the development, we can add any suggestion base on your input.

**Comment by mr-tron** [comment]
You can remove register. It`s deprecated function.

File: src/test/java/com/venafi/vcert/sdk/connectors/tpp/TppConnectorIT.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+        assertThat(zoneConfiguration).isNotNull();
+        assertThat(zoneConfiguration.organization()).isNull();
+        assertThat(zoneConfiguration.organizationalUnit()).isNotNull();
+        assertThat(zoneConfiguration.organizationalUnit()).isEmpty();
+        assertThat(zoneConfiguration.country()).isNull();
+        assertThat(zoneConfiguration.province()).isNull();
+        assertThat(zoneConfiguration.locality()).isNull();
+        assertThat(zoneConfiguration.policy()).isNotNull();
+        assertThat(zoneConfiguration.policy().subjectCNRegexes()).containsExactly(".*");
```
Comment (mr-tron) [inline src/test/java/com/venafi/vcert/sdk/connectors/tpp/TppConnectorIT.java]: i don't know how and what type of regular expressions dialect do you use, but remember, that regexp should match full string. so for example regexp `.*\.example\.com` should doesn't match cn `www.example.company`

**Comment by mr-tron** [comment]
Also please move hardcoded tokens and urls in tests to environment variables.

**Comment by pegerto** [comment]
> You can remove register. It`s deprecated function.

Done

File: src/test/java/com/venafi/vcert/sdk/connectors/tpp/TppConnectorIT.java (hunk 1 above)
Comment (pegerto) [reply src/test/java/com/venafi/vcert/sdk/connectors/tpp/TppConnectorIT.java]: Hello I see, in this case, the test is validating that the API requires its correct from a stub external service https://github.com/opencredo/vcert-java-1/blob/68f1d07aaf4fc61e613361f68fb43b5932712cea/src/test/resources/mappings/cloud.certificatepolicies.di.json#L40

**Comment by mr-tron** [comment]
As I understand you just rewrite golang code to java. We are going to make some changes in the nearest future. May be you can just write better code from beginning and not copy our errors.
```
    certificate.Request has unused public attribute PublicKeyAlgorithm - should be removed (duplicated by KeyType)
    certificate.Request has public field CSR []byte what make misunderstoonding what can bein this field: pem or asn encoded csr. Should be replaced with private filed csr and few methods like SetCSR (may be with auto detect type), GetCSRasn1, GetCSRpem
    few methods accept empty interface as private key. should be replaced with crypto.Signer interface
    rewrite retrive function for cloud connector. should be getting certID by cn/thumbprint/pickupid -> getting certificate by certID via /v1/certificate/ {id}
    `func (cfg *Config) LoadFromFile() error` should be replaced with `func LoadConfigFromFile(path) (*Config, error)`
    SetBaseURL should be replaced with additional argument in NewConnector function //discuss
```

**Comment by pegerto** [comment]
Hello @m-tron

I fully understand we discuss this in our proposal,  it will be good to have a more fluent API and refactor sections on our code, unfortunately, we are time constrained, so we implement as match parity as possible for tracking.

We will try to implement part of the refactoring in this initial PR.

File: src/test/java/com/venafi/vcert/sdk/connectors/tpp/TppConnectorIT.java (hunk 1 above)
Comment (mr-tron) [reply src/test/java/com/venafi/vcert/sdk/connectors/tpp/TppConnectorIT.java]: I don`t big specialist in java. I did`t know what java `pattern.matches` which match whole string. In golang you need to add `^` and `$` to your regexp for matching whole string.

**Comment by pegerto** [comment]
> Also please move hardcoded tokens and urls in tests to environment variables.

we have externalized acceptance test so it can be run by an integration environment and document the variables in the README file

**Comment by mr-tron** [comment]
Can you also add example of library usages like https://github.com/Venafi/vcert/blob/master/example/main.go with 
1. enrolling certificate
2. retrieving certificate
3. renew
4. checking certificate request with zone policies (I think CN will be enough)

**Comment by pegerto** [comment]
@mr-tron an Example class has been added this PR, The recommendation is to create a different project for examples, or submodule the project if more examples are required.

**Pull request merged**